File: JavaPartI/la_mergeSort.java
public class Solution {
  public int[] mergeSort(int[] array) {
    // Write your solution here.
        if (array == null){
	     return array;
	}
	int[] helper = new int[array.length];

	mergeSort(array, helper, 0, array.length-1);
	return array;
    
  }
  private void mergeSort(int[] array, int[] helper, int left, int right){
	if (left>=right){
		return;
	}
	int mid = left + (right - left)/2;
	mergeSort(array, helper, left, mid);
	mergeSort(array, helper, mid+1, right);
	merge(array, helper, left, mid, right);
		
  }
  private void merge(int[] array, int[] helper, int left, int mid, int right){
      
      for (int i = left; i<= right; i++){
      	helper[i] = array[i];
      }
      int pt1 = left;
      int pt2 = mid + 1;
      while(pt1 <= mid && pt2 <= right){
      	if (helper[pt1] <= helper[pt2]){
      		array[left++] = helper[pt1++];
      	}
      	else {
      		array[left++] = helper[pt2++];
      	}
      }
      while (pt1<=mid){
      	array[left++] = helper[pt1++];
      }
  
  }
}


/// way 2 (recursion with return value)
public class Solution {
  public int[] mergeSort(int[] array) {
    // Write your solution here.
    if (array == null || array.length == 0) {
      return array;
    }
    return mergeHelper(array, 0, array.length - 1);
  }
  
  
  private int[] mergeHelper(int[] array, int left, int right) {
    if (left == right) {
      return new int[]{array[left]};
    }
    int mid = left + (right - left) / 2;
    int[] arr1 = mergeHelper(array, left, mid);
    int[] arr2 = mergeHelper(array, mid + 1, right);
    return merge(arr1, arr2);
  }
  
  private int[] merge(int[] arr1, int[] arr2) {
    int len1 = arr1.length, len2 = arr2.length;
    int i = 0, j = 0, k = 0;
    int[] res = new int[len1 + len2];
    while (i < len1 && j < len2) {
      if (arr1[i] < arr2[j]) {
        res[k++] = arr1[i++];
      } else {
        res[k++] = arr2[j++];
      }
    }
    while (i != len1) {
      res[k++] = arr1[i++];
      
    }
    while (j != len2) {
      res[k++] = arr2[j++];
    }
    return res;
  }
}

File: pythonPart/la_mergeSort.py
#### without return value

class Solution(object):
  def mergeSort(self, array):
    """
    input: int[] array
    return: int[]
    """
    # write your solution here
    
    if array is None or len(array) == 0:
      return array
    copy = [0] * len(array)
    self.mergeHelper(array, copy, 0, len(array) - 1)
    return array
  
  def mergeHelper(self, array, copy, i, j):
    if i == j:
      return
    m = i + (j - i) // 2
    self.mergeHelper(array, copy, i, m)
    self.mergeHelper(array, copy, m + 1, j)
    self.merge(array, copy, i, m, j)
    
    
  def merge(self, array, helper, left, m, right):
    for i in range(left, right + 1):
      helper[i] = array[i]
    i = left
    j = m + 1
    k = left
    while i <= m and j <= right:
      if helper[i] < helper[j]:
        array[k] = helper[i]
        i += 1
      else:
        array[k] = helper[j]
        j += 1
      k += 1
    while i <= m:
      array[k] = helper[i]
      k += 1
      i += 1


### with return value
class Solution(object):  
  def mergeSort(self, array):
    """
    input: int[] array
    return: int[]
    """
    # write your solution here
    
    if array is None or len(array) == 0:
      return array
    return self.mergeHelper(array, 0, len(array) - 1)
   
  
  def mergeHelper(self, array, i, j):
    if i == j:
      out = [array[i]]
      return out
    m = i + (j - i) // 2
    leftseg = self.mergeHelper(array, i, m)
    rightseg = self.mergeHelper(array, m + 1, j)
    out = self.merge(leftseg, rightseg)
    return out
    
  def merge(self, left, right):
    n1 = len(left)
    n2 = len(right)
    n = n1 + n2
    out = [0] * n
    i, j = 0, 0
    k = 0
    while i < n1 and j < n2:
      if left[i] < right[j]:
        out[k] = left[i]
        i += 1
      else:
        out[k] = right[j]
        j += 1
      k += 1
    while j < n2:
      out[k] = right[j]
      k += 1
      j += 1
    while i < n1:
      out[k] = left[i]
      i += 1
      k += 1
    return out
